### DISCUSSION TITLE
Confused by handling inputRef (or missing ref). Suggests for an improvement.

### DISCUSSION THREAD

**Discussion by leochoo** [react-hook-form/react-hook-form#7350] [Q&A] [answered]
# TL; DR
1. The docs for handling inputRef does not seem comprehensive enough.
2. I could only get my code working by doing in a way that is not shown in the referred docs.
3. I would highly appreciate working examples and clearer explanations.

# Issue

These two documents explain how to work with `innerRef`,`inputRef` and handle "Missing `ref`".

Let me refer to theese  as `doc1` and `doc2`
doc1. https://react-hook-form.com/api/useform/register#registerRef
doc2. https://www.react-hook-form.com/migrate-v6-to-v7/

But it was hard to get my code working following the above two docs.

1. The example that should NOT work actualy WORKS!
In `doc1`, it says
```js
// not working, because ref is not assigned
<TextInput {...register('test')} />
```

But when I tried this on my code like below, it works.
```js
<TextField
          className={styles.formInput}
          label="Email"
          type="email"
          variant="outlined"
          error={!!errors.email}
          inputProps={{ "aria-label": "email" }}
          {...register("email", {
            required: {
              value: true,
              message: "Email Address is required.",
            },
            maxLength: {
              value: 48,
              message: "Email Address must be shorter than 48 letters.",
            },
          })}
          helperText={errors.email?.message}
        />
```
MUI TextField only exposes inputRef hence I should be using `inputRef` but doing it the "wrong way" works.
Could you please help me understand why this would work in this case and why this shouldn't be the correct way?

2. "correct way" does not seem to be documented

Following the two examples from the docs above under "Missing ref" and "How to work with innerRef, inputRef?"
I changed my code above to this below.
```js
  const name = register("name", {
    required: {
      value: true,
      message: "Full Name is required.",
    },
    maxLength: {
      value: 48,
      message: "Full Name must be shorter than 48 letters.",
    },
  });
....

        <TextField
          // omitted unrelated lines
          inputProps={{ "aria-label": "name" }}
          inputRef={name.ref}
        />
```

However, this does not seem to register correctly and is causing a `null` error.

Finally after looking at [StackOverflow answer](https://stackoverflow.com/a/67015773/7414387), I was able to get the code working like this:
```js
  const { ref: nameRef, ...inputProps } = register("name", {
  // omitted unchanged lines
  });
....
        <TextField
          // omitted unchanged lines
          inputProps={{ "aria-label": "name" }}
          {...inputProps}
          inputRef={nameRef}
        />
```

# What confused me
1.  Being able to set up a name for a ref like this `const { ref: nameRef, ...inputProps } ` => I could not find the related info in the doc1 or doc2
2. The need to pass inputProps like `{...inputProps}` => I could not find the related info in the doc1 or doc2.
3. I don't think the above docs provide me enough information or guide to related docs to solve the issue.

# What I want to ask here
I would like to know what was the intended or correct way to handle this
and also more working examples in the docs.

Thanks for your work always.

**Comment by bluebill1049** [comment, marked as answer]
Thanks for the detailed discussion and feedback above. to be completely honest with you (and everyone who reads this). This is one of the most annoying parts of dealing with the form input with uncontrolled inputs because every library named their ref props differently. I can name a few of them:

- ref (which is standard)
- ref exposed but not the actual input
- inputRef
- innerRef
- no ref forward at all

which is leads to a difficult topic for uncontrolled form library like react hook form to address except bring more config and that's a double-edged sword as well... it has the potential to resolve and address some of the problems which you have raised above but it potential increase even more confusion with unknow usage (which is the reason why such proposal is not going ahead). relevant feature request: #4629

### What's next?

So is this a problem for react hook form to resolve? To be honest, I hope not unless we all have a standard way to expose input ref, and adding another config inside the register is something I prefer to avoid (considering there are already quite a few configs).

I would recommend using `Controller/useController` for that external controlled component that doesn't expose standard `ref`, and address some of the concerns in the documentation to lead or at least info the user what's the correct usage.

In addition, if you/anyone has a better strategy for addressing this issue, the documentation is open source as well, so feel free to send us a PR to address some of the concerns which you have and help everyone.

> "correct way" does not seem to be documented

not the best, but we have a section under the register API, let us know or send us a PR so we can improve on it.

![Screen Shot 2021-12-23 at 5 59 56 pm](https://user-images.githubusercontent.com/10513364/147201416-e1519611-b66e-407c-8aa2-adcb141b58c4.png)
